packages/nango-yaml/lib/parser.v2.ts

```diff
@@ -0,0 +1,261 @@
+            this.warnings.push(...this.modelsParser.warnings);
+        }
+
+        for (const integrationName in yaml.integrations) {
+            const integration = yaml.integrations[integrationName];
+            if (!integration) {
+                continue;
+            }
+
+            // check that models are used only once per integration
+            const usedModels = new Set<string>();
+
+            const syncs = integration['syncs'];
+            const actions = integration['actions'];
+            const postConnectionScripts: string[] = integration['post-connection-scripts'] || [];
+
+            const parsedSyncs = this.parseSyncs({ syncs, usedModels, integrationName });
+            const parseActions = this.parseActions({ actions, usedModels, integrationName });
+
+            const parsedIntegration: NangoYamlParsedIntegration = {
+                providerConfigKey: integrationName,
+                syncs: parsedSyncs,
+                actions: parseActions,
+                postConnectionScripts
+            };
+
+            output.push(parsedIntegration);
+        }
+
+        this.parsed = {
+            yamlVersion: 'v2',
+            integrations: output,
+            models: this.modelsParser.parsed
+        };
+        return this.errors.length <= 0;
+    }
+
+    parseSyncs({
+        syncs,
+        usedModels,
+        integrationName
+    }: {
+        syncs: NangoYamlV2Integration['syncs'];
+        usedModels: Set<string>;
+        integrationName: string;
+    }): ParsedNangoSync[] {
+        const parsedSyncs: ParsedNangoSync[] = [];
+
+        for (const syncName in syncs) {
+            const sync = syncs[syncName];
+            if (!sync) {
+                continue;
+            }
+
+            const modelNames = new Set<string>();
+
+            const modelOutput = this.getModelForOutput({ rawOutput: sync.output, usedModels, name: syncName, type: 'sync', integrationName });
+            if (!modelOutput) {
+                continue;
+            }
+            modelOutput.forEach((m) => modelNames.add(m.name));
+
+            const modelInput = this.getModelForInput({ usedModels, rawInput: sync.input, name: syncName, type: 'sync', integrationName });
+            if (modelInput) {
+                modelNames.add(modelInput.name);
+            }
+
+            const endpoints: NangoSyncEndpoint[] = [];
+            if (sync.endpoint) {
+                const tmp = Array.isArray(sync.endpoint) ? sync.endpoint : [sync.endpoint];
+
+                if (tmp.length !== modelOutput.length) {
+                    this.errors.push(new ParserErrorEndpointsMismatch({ syncName: syncName, path: [integrationName, 'syncs', syncName] }));
+                    continue;
+                }
+
+                let failed = false;
+                for (const endpoint of tmp) {
+                    const split = endpoint.split(' ') as [HTTP_VERB, string];
+
+                    if (this.endpoints.has(endpoint)) {
```

Comment:
linked to my question above about the `endpoints` class field: this function relies on the internal state of the parser instance, which might be different if the parsing functions are reordered. Would it be simpler to first deal with the pure parsing and then in a second pass, take care of the validations like duplicate endpoints, etc... it would make those parsing function smaller, without side effects and easier to compose/reorder

Reply:
fair